vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go

```diff
@@ -0,0 +1,395 @@
+	// Required.
+	DeletionPolicy DeletionPolicy `json:"deletionPolicy" protobuf:"bytes,2,opt,name=deletionPolicy"`
+
+	// driver is the name of the CSI driver used to create the physical snapshot on
+	// the underlying storage system.
+	// This MUST be the same as the name returned by the CSI GetPluginName() call for
+	// that driver.
+	// Required.
+	Driver string `json:"driver" protobuf:"bytes,3,opt,name=driver"`
+
+	// name of the VolumeSnapshotClass to which this snapshot belongs.
+	// +optional
+	VolumeSnapshotClassName *string `json:"volumeSnapshotClassName,omitempty" protobuf:"bytes,4,opt,name=volumeSnapshotClassName"`
+
+	// source specifies from where a snapshot will be created.
+	// This field is immutable after creation.
+	// Required.
+	Source VolumeSnapshotContentSource `json:"source" protobuf:"bytes,5,opt,name=source"`
+}
+
+// VolumeSnapshotContentSource represents the CSI source of a snapshot.
+// Exactly one of its members must be set.
+// Members in VolumeSnapshotContentSource are immutable.
+// TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotContentSource members
+// will be immutable once specified.
+type VolumeSnapshotContentSource struct {
+	// volumeHandle specifies the CSI "volume_id" of the volume from which a snapshot
+	// should be dynamically taken from.
+	// This field is immutable.
+	// +optional
+	VolumeHandle *string `json:"volumeHandle,omitempty" protobuf:"bytes,1,opt,name=volumeHandle"`
+
+	// snapshotHandle specifies the CSI "snapshot_id" of a pre-existing snapshot on
+	// the underlying storage system.
+	// This field is immutable.
+	// +optional
+	SnapshotHandle *string `json:"snapshotHandle,omitempty" protobuf:"bytes,2,opt,name=snapshotHandle"`
+}
+
+// VolumeSnapshotContentStatus is the status of a VolumeSnapshotContent object
+type VolumeSnapshotContentStatus struct {
+	// snapshotHandle is the CSI "snapshot_id" of a snapshot on the underlying storage system.
+	// If not specified, it indicates that dynamic snapshot creation has either failed
+	// or it is still in progress.
+	// +optional
+	SnapshotHandle *string `json:"snapshotHandle,omitempty" protobuf:"bytes,1,opt,name=snapshotHandle"`
+
+	// creationTime is the timestamp when the point-in-time snapshot is taken
+	// by the underlying storage system.
+	// In dynamic snapshot creation case, this field will be filled in with the
+	// "creation_time" value returned from CSI "CreateSnapshotRequest" gRPC call.
+	// For a pre-existing snapshot, this field will be filled with the "creation_time"
+	// value returned from the CSI "ListSnapshots" gRPC call if the driver supports it.
+	// If not specified, it indicates the creation time is unknown.
+	// The format of this field is a Unix nanoseconds time encoded as an int64.
+	// On Unix, the command `date +%s%N` returns the current time in nanoseconds
+	// since 1970-01-01 00:00:00 UTC.
+	// +optional
+	CreationTime *int64 `json:"creationTime,omitempty" protobuf:"varint,2,opt,name=creationTime"`
```

Comment:
We've had this discussion before, but I want to revisit it one last time. Why do we need `CreationTime`, `RestoreSize`, `ReadyToUse`, and `Error` duplicated on both the `VolumeSnapshot` and the `VolumeSnapshotContent` objects?

If we agree the duplication makes sense, please make sure to apply all the comments I left for these objects on the `VolumeSnapshot` side here as well.

Reply:
TODO: Document why we have them in both objects. 
- status on content can be used for filtering when importing a volumesnapshot.
- volumsnapshot status is used by end users because they can't see volumesnapshotcontent
- Add how they are set and how they are synced.  Eventual consistency.

Reply:
Done

Reply:
Done